File: LeetCode/[M] - 1466. Reorder Routes to Make All Paths Lead to the City Zero/solution.py
import collections
from typing import List


class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        graph = collections.defaultdict(set)
        paths = collections.defaultdict(set)
        for x, y in connections:
            paths[x].add(y)
            graph[x].add(y)
            graph[y].add(x)

        cnt = 0

        visited = set()
        stack = [0]
        while stack:
            v = stack.pop()
            visited.add(v)
            for w in graph[v]:
                if w not in visited:
                    stack.append(w)
                    if w in paths[v]:
                        cnt += 1
        return cnt

File: LeetCode/[M] - 1466. Reorder Routes to Make All Paths Lead to the City Zero/solution.go
package main

func minReorder(n int, connections [][]int) int {
	graph := make(map[int]map[int]struct{})
	order := make(map[int]map[int]struct{})
	for i := 0; i < n; i++ {
		graph[i] = make(map[int]struct{})
		order[i] = make(map[int]struct{})
	}
	for _, conn := range connections {
		x, y := conn[0], conn[1]
		order[x][y] = struct{}{}
		graph[x][y] = struct{}{}
		graph[y][x] = struct{}{}
	}

	cnt := 0

	visited := make(map[int]struct{})
	stack := []int{0}
	for len(stack) != 0 {
		v := stack[0]
		stack = stack[1:]
		visited[v] = struct{}{}
		for w := range graph[v] {
			if _, ok := visited[w]; !ok {
				stack = append(stack, w)
				if _, ok := order[v][w]; ok {
					cnt++
				}
			}
		}
	}
	return cnt
}
